File: thesis/algorithms/java/p.java
import java.io.*;
import java.util.*;

public class P {
    public static void mergeSort(int[] arr, int l, int r) {
        if (l < r) {
            int m = (l + r) / 2;
            mergeSort(arr, l, m);
            mergeSort(arr, m + 1, r);
            merge(arr, l, m, r);
        }
    }

    public static void merge(int[] arr, int l, int m, int r) {
        int[] left = Arrays.copyOfRange(arr, l, m + 1);
        int[] right = Arrays.copyOfRange(arr, m + 1, r + 1);
        int i = 0, j = 0, k = l;

        while (i < left.length && j < right.length) {
            arr[k++] = (left[i] <= right[j]) ? left[i++] : right[j++];
        }
        while (i < left.length) arr[k++] = left[i++];
        while (j < right.length) arr[k++] = right[j++];
    }

    public static int[] readJsonArray(String path) throws IOException {
        String content = new String(Files.readAllBytes(new File(path).toPath()));
        content = content.replace("[", "").replace("]", "").trim();
        String[] parts = content.split(",");
        return Arrays.stream(parts).mapToInt(Integer::parseInt).toArray();
    }

    public static void main(String[] args) throws IOException {
        String size = args.length > 0 ? args[0] : "small";
        String path = "datasets/" + size + "/merge_sort.json";
        int[] arr = readJsonArray(path);
        mergeSort(arr, 0, arr.length - 1);
        System.out.println("Ordenado " + arr.length + " elementos (" + size + ")");
    }
}

File: thesis/algorithms/kotlin/p.kt
import java.io.File

fun mergeSort(arr: List<Int>): List<Int> {
    if (arr.size <= 1) return arr
    val middle = arr.size / 2
    val left = mergeSort(arr.subList(0, middle))
    val right = mergeSort(arr.subList(middle, arr.size))
    return merge(left, right)
}

fun merge(left: List<Int>, right: List<Int>): List<Int> {
    val result = mutableListOf<Int>()
    var i = 0
    var j = 0
    while (i < left.size && j < right.size) {
        if (left[i] <= right[j]) {
            result.add(left[i])
            i++
        } else {
            result.add(right[j])
            j++
        }
    }
    result.addAll(left.drop(i))
    result.addAll(right.drop(j))
    return result
}

fun main(args: Array<String>) {
    val size = if (args.isNotEmpty()) args[0] else "small"
    val path = "datasets/$size/merge_sort.json"

    val content = File(path).readText()
        .replace("[", "")
        .replace("]", "")
        .trim()

    val arr = if (content.isNotEmpty()) content.split(",").map { it.trim().toInt() } else listOf()
    val sorted = mergeSort(arr)

    println("Ordenado ${sorted.size} elementos ($size)")
}
